app/com/arpnetworking/metrics/portal/scheduling/JobExecutorActor.java

```diff
@@ -96,42 +100,108 @@ public void preStart() throws Exception {
+        if (!_cachedJob.isPresent()) {
+            LOGGER.info()
+                    .setMessage("initializing")
+                    .addData("ref", ref)
+                    .log();
+            _cachedJob = Optional.of(new CachedJob<>(_injector, ref));
+        }
+
+        final JobRef<T> oldRef = _cachedJob.get().getRef();
+        if (!oldRef.equals(ref)) {
+            LOGGER.error()
+                    .setMessage("JobRef in received message does not match cached JobRef")
+                    .addData("cached", oldRef)
+                    .addData("new", ref)
+                    .log();
+            killSelf();
+            throw new IllegalStateException(String.format("got JobRef %s, but already initialized with %s", ref, oldRef));
         }
+
+        return _cachedJob.get();
     }
 
-    private void scheduleNextTick(final Instant wakeUpBy) {
-        timers().startSingleTimer("TICK", Tick.INSTANCE, timeUntilNextTick(wakeUpBy));
+
+    private JobRef<T> unsafeJobRefCast(@SuppressWarnings("rawtypes") final JobRef ref) {
+        // THIS MAKES ME SO SAD. But there's simply no way to plumb the type information through Akka.
+        @SuppressWarnings("unchecked")
+        final JobRef<T> typedRef = ref;
+        return typedRef;
     }
 
-    private void executeAndScheduleNextTick(
-            final JobRepository<T> repo,
-            final Organization org,
-            final Job<T> job,
-            final Instant scheduled
-    ) {
-        final Optional<Instant> nextScheduled = job.getSchedule().nextRun(Optional.of(scheduled));
-        repo.jobStarted(job.getId(), org, scheduled);
+    private void attemptExecuteAndUpdateRepository(final Instant scheduled) {
+        if (!_cachedJob.isPresent()) {
+            LOGGER.error()
+                    .setMessage("unable to execute: executor is not initialized")
+                    .log();
+            return;
+        }
+
+        final CachedJob<T> cachedJob = _cachedJob.get();
+
+        final JobRef<T> ref = cachedJob.getRef();
+        final Job<T> job = cachedJob.getJob();
+        final JobRepository<T> repo = ref.getRepository(_injector);
+
+        if (_currentlyExecuting.getAndSet(true)) {
+            return;
+        }
+
+        try {
+            repo.jobStarted(ref.getJobId(), ref.getOrganization(), scheduled);
+        } catch (final NoSuchElementException error) {
+            _currentlyExecuting.set(false);
+            killSelf();
+            return;
+        }
+
         job.execute(getSelf(), scheduled)
                 .handle((result, error) -> {
-                    if (error == null) {
-                        repo.jobSucceeded(job.getId(), org, scheduled, result);
-                    } else {
-                        repo.jobFailed(job.getId(), org, scheduled, error);
+                    try {
+                        if (error == null) {
+                            repo.jobSucceeded(ref.getJobId(), ref.getOrganization(), scheduled, result);
```

Comment:
IMO the succeeded/failed branches should log some debug/info (or at least trace) level data about the change in state

Reply:
What about having metrics here?